Dashboard v2: User Profile - Public Gravatar section

brunobasto opened the pull request:
Fixes https://linear.app/a8c/issue/DOTDASH-353/implement-profile-public-gravatar-profile-section

⚠️ This depends on https://github.com/Automattic/wp-calypso/pull/105582 for validations to work properly.

## Proposed Changes

This is a follow-up on https://github.com/Automattic/wp-calypso/pull/105491

The "Public Gravatar profile" section was already mostly functional. On this PR I've added:
- Form validations
- Snackbar notifications
- Unit tests

## Why are these changes being made?

To implement the proposed designs for the new account profile page.

## Testing Instructions

<img width="1320" height="1128" alt="Calypso Profile (3)" src="https://github.com/user-attachments/assets/9c031939-0fea-4fc5-8bbd-4076887401c3" />

1. Navigate to http://calypso.localhost:3000/v2/me/profile
2. Type an invalid URL in the `Web address` field.
3. Verify that the validation works:

<img width="612" height="87" alt="Calypso Profile (1)" src="https://github.com/user-attachments/assets/f628304c-9485-44f7-b8e8-639ff830062c" />

4. Type a display name longer than 250 characters
5. Verify that the validation works:

<img width="612" height="87" alt="Calypso Profile (2)" src="https://github.com/user-attachments/assets/355214ab-4ed6-470a-8db3-1604dfff0c42" />

6. Fix the validation errors
7. Submit the form
8. Verify that the snackbar notification shows up:

<img width="989" height="660" alt="Screenshot 2025-09-08 at 11 05 36" src="https://github.com/user-attachments/assets/e6c05394-37fd-4a09-8b14-8769bec949aa" />

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

brunobasto committed:
Add form validations

brunobasto committed:
Add snackbar notifications

brunobasto committed:
Add tests for the notifications

abdullah-kasim commented on client/dashboard/me/profile-gravatar/test/notifications.test.ts:
```diff
@@ -0,0 +1,154 @@
+				{ type: 'snackbar' }
+			);
+			expect( mockCreateSuccessNotice ).toHaveBeenCalledTimes( 1 );
+		} );
```
This test currently defines the `onSuccess()` function and then calls it. It's basically adding business logic to the test, and then testing the very same business logic in the same test. Which, in a less nice way, it basically does not test the component at all. Maybe you intended to change it later? 

If not, then I'd suggest that we change it to something that renders the component, mock the API call, submits the form, and see if `mockCreateSuccessNotice` gets called.

brunobasto replied:
Thanks for reviewing @abdullah-kasim. I've updated the tests in https://github.com/Automattic/wp-calypso/pull/105605/commits/d7a882058169b0fc2d147fdda3fdce70220a0f60

Let me know if that looks better.

abdullah-kasim replied:
Yep first glance that looks better 👍 .

abdullah-kasim commented on client/dashboard/me/profile-gravatar/test/notifications.test.ts:
```diff
@@ -0,0 +1,154 @@
+
+			const callArgs = mockCreateSuccessNotice.mock.calls[ 0 ];
+			expect( callArgs[ 1 ] ).toEqual( { type: 'snackbar' } );
+		} );
```
Similar issue as per previous commit.

brunobasto replied:
I'll resolve this one to keep the conversation on the main thread at https://github.com/Automattic/wp-calypso/pull/105605#discussion_r2328921348

Also, I've updated the tests on https://github.com/Automattic/wp-calypso/commit/d7a882058169b0fc2d147fdda3fdce70220a0f60

abdullah-kasim commented on client/dashboard/me/profile-gravatar/test/notifications.test.ts:
```diff
@@ -0,0 +1,154 @@
+
+			expect( mockCreateErrorNotice ).toHaveBeenCalledWith( errorMessage, { type: 'snackbar' } );
+			expect( mockCreateErrorNotice ).toHaveBeenCalledTimes( 1 );
+		} );
```
Similar issue as previous comment. There's a few more down there but I'll stop here.

brunobasto replied:
I'll resolve this one to keep the conversation on the main thread at https://github.com/Automattic/wp-calypso/pull/105605#discussion_r2328921348

Also, I've updated the tests on https://github.com/Automattic/wp-calypso/commit/d7a882058169b0fc2d147fdda3fdce70220a0f60

abdullah-kasim commented on client/dashboard/me/profile-gravatar/test/validation.test.ts:
```diff
@@ -0,0 +1,72 @@
+		it( 'should require display name', () => {
+			expect( validateDisplayName( '' ) ).toBe( 'Display name is required.' );
+			expect( validateDisplayName( '   ' ) ).toBe( 'Display name is required.' );
+		} );
```
Similar issue - the test is writing its own business logic.

brunobasto replied:
I'll resolve this one to keep the conversation on the main thread at https://github.com/Automattic/wp-calypso/pull/105605#discussion_r2328921348

Also, I've updated the tests on https://github.com/Automattic/wp-calypso/commit/d7a882058169b0fc2d147fdda3fdce70220a0f60

andrea-sdl reviewed (commented):
@brunobasto I was comparing this to the Design and I think there might be 2 things to implement, let me know what you think

1. [Use the description field on the Section Header](https://github.com/Automattic/wp-calypso/blob/7d133c054295bc8172cc495159f64007ffb58951/client/dashboard/me/profile-gravatar/index.tsx#L152) instead of printing the Text Separately (this will make the description match the UI)
2. Keep the avatar size at 14px (This is more a guess of mine, but looking at the UI I wouldn't expect the circle to grow once you add the avatar)

Here's the UI with my notes, in case it helps.

<img width="1528" height="984" alt="Image" src="https://github.com/user-attachments/assets/2d840ad7-e7f5-4d0e-8438-2be3fc245935" />

brunobasto committed:
Use `SectionHeader` description

brunobasto committed:
Adjust avatar size to match mockups

brunobasto commented:
> @brunobasto I was comparing this to the Design and I think there might be 2 things to implement, let me know what you think
> 
> 1. [Use the description field on the Section Header](https://github.com/Automattic/wp-calypso/blob/7d133c054295bc8172cc495159f64007ffb58951/client/dashboard/me/profile-gravatar/index.tsx#L152) instead of printing the Text Separately (this will make the description match the UI)

Thank you for reviewing @andrea-sdl! Good catch on using the description prop of the `<SectionHeader>` component. Updated in https://github.com/Automattic/wp-calypso/pull/105605/commits/71437b067e38a05d4b1b0683233e6b82fe167f66

> 2. Keep the avatar size at 14px (This is more a guess of mine, but looking at the UI I wouldn't expect the circle to grow once you add the avatar)

The `<EditGravatar>` component was something that already existed prior to my changes and I wasn't sure if I was supposed to reduce its size - I was afraid the smaller size would make it harder to see details on the image - but I asked on slack and Lucas confirmed the smaller size is what we should use, so I updated it on https://github.com/Automattic/wp-calypso/pull/105605/commits/fa0fa9a1e790c89cc23184808ba245bbd3f3924d

brunobasto committed:
Improve tests to test actual component behavior

andrea-sdl commented on client/dashboard/me/profile-gravatar/test/notifications.test.tsx:
```diff
@@ -0,0 +1,154 @@
+} ) );
+
+// Mock URL validation
+jest.mock( '../../../../lib/importer/url-validation', () => ( {
```
@brunobasto what's the benefit of mocking the `url-validation`? Is there something that breaks at the test level?

brunobasto replied:
right, no real benefit. I was gonna change this before pushing, but forgot. I'll update

brunobasto replied:
Removed in https://github.com/Automattic/wp-calypso/pull/105605/commits/22969ec582e1ede786ea7ecd70862cf49d4fb284

andrea-sdl commented on client/dashboard/me/profile-gravatar/test/notifications.test.tsx:
```diff
@@ -0,0 +1,154 @@
+
+// Mock the profile mutation
+const mockMutate = jest.fn();
+jest.mock( '@automattic/api-queries', () => ( {
```
@brunobasto I wonder if we could skip the mocking and use `nock` instead  (example https://github.com/Automattic/wp-calypso/blob/d7a882058169b0fc2d147fdda3fdce70220a0f60/client/dashboard/me/preferences-language/test/index.test.tsx#L13 ) This way we won't hide the whole stack to the testing but only the HTTP calls to the endpoints.

brunobasto replied:
Good idea. Updated in https://github.com/Automattic/wp-calypso/pull/105605/commits/363f107cf0c34631ba41d19aa7c021b34fa746cb

brunobasto committed:
Remove unnecessary mocks

brunobasto committed:
Using `nock` instead of mocking the mutations

arthur791004 commented on package.json:
```diff
@@ -135,7 +135,7 @@
 		"translate:dependencies": "rm -rf build/dependency-strings && mkdir -p build/strings && wp-babel-makepot './node_modules/@wordpress/**/build-module/**/*.js' --ignore '**/*.native.js' --base './' --dir './build/dependency-strings' --output './build/strings/calypso-dependencies.pot'",
 		"translate:calypso": "wp-babel-makepot './{client,packages,apps}/**/*.{js,jsx,ts,tsx}' --ignore '**/node_modules/**,**/test/**,**/*.d.ts' --base './' --dir './build/strings' --output './public/calypso-strings.pot'",
-		"tsc": "NODE_OPTIONS='--max-old-space-size=4096' tsc",
+		"tsc": "NODE_OPTIONS='--max-old-space-size=8192' tsc",
```
Let's revert this and see how https://github.com/Automattic/wp-calypso/pull/105638 goes 🙂

brunobasto replied:
Makes sense. done

andrea-sdl commented on client/dashboard/me/profile-gravatar/test/validation.test.tsx:
```diff
@@ -0,0 +1,125 @@
+import '@testing-library/jest-dom';
+import { render } from '../../../test-utils';
+import GravatarProfileSection from '../index';
+import type { UserProfile } from '@automattic/api-core';
```
I think this needs to be updated to use `UserSettings` and the `MockProfile` to use a `Partial<UserSettings>` or the tests won't pass.

brunobasto replied:
ah, thank you! I haven't seen this

brunobasto committed:
Fix unit tests - use new UserSettings type

brunobasto committed:
Properly define mock

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/20098767

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @brunobasto for including a screenshot in the description! This is really helpful for our translators.
